### ISSUE TITLE
VS runner hangs on Run All

### ISSUE THREAD

**Issue by svickers** [xunit/xunit#611]
When selecting Run All from test explorer this is shown in the output and the tests never execute:

------ Run test started ------
[xUnit.net 00:00:00.3441381]   Discovering: ConsoleApplication1
[xUnit.net 00:00:00.6192942]   Discovered:  ConsoleApplication1

If I debug an individual test, or Run > Failed, Not Run, or Passed everything works fine.  VS2015 Community edition.

**Comment by TheFastCat** [comment]
I am also experiencing this

**Comment by kkm000** [comment]
cf. nunit/nunit3-vs-adapter#51 and nunit/nunit3-vs-adapter#54 -- maybe a common VS issue? Suspiciously similar.

**Comment by WaffleSquirrel** [comment]
If you are running Visual Studio on a 64-bit machine, check this setting in Visual Studio. From the top menu, go to Test > Test Settings > Default Processor Architecture. 

Change that setting to X64 if it isn't already. The custom test runners internally utilize some loading/unloading of app domains, so sometimes there's a conflict under the covers using the default 32-bit target. Yep, similar thing with NUnit as mentioned above.

I don't have the issue on Windows 10 VS2015. But I had the issue on a different machine with Windows 7 VS2015.

Hope this helps.

**Comment by bradwilson** [comment]
I believe the major difference is that, in VS 2015, they re-wrote the test runner, but only for 32-bit tests. When you want to run in 64-bit mode, it's using the older test runner from VS 2013 (which appears to be more stable).

**Comment by TheFastCat** [comment]
This change does not resolve my issue - it seems like an async issue? When "Run All" is selected four tests will execute, complete and then the runner will "hang" - same with any selected subset of tests and "Run selected tests". This is Visual Studio 2013 with architecture X64

**Comment by kkm000** [comment]
@bradwilson: Thanks, I'll try. But generally, do you think is this a Visual Studio problem, and can it be worked around somehow in the xUnit runner? I think that NUnit has attempted [workaround by running unload in a thread](https://sourceforge.net/p/nunit/bugs/735/) (long before VS2015!). Apparently, the problem has been long lingering there, as @TheFastCat notes. Also, NUnit fails slightly "better": the test results are reported just fine, only the test run never completes, but can be stopped with the Cancel button. The tests do change to red or green. This is unlike xUnit, which hangs before reporting the results, apparently because its "main" (in some sense) thread hangs. I am not advocating implementing same type of a semi-workaround, just hope it would provide some insight into the problem.

Another observation is that at the time of the deadlock, there is also a thread pool thread blocked in a COM call, but that's in the MTA of the process, and I think it might not be significant.

Is there any way to debug this type of a deadlock? I can reproduce it reliably, and I would be glad to test a solution if you could think of one. It would be amazing to get to the bottom of it!

**Comment by svickers** [comment]
Default architecture recommendation from @WaffleSquirrel fixed it for me.

**Comment by onovotny** [comment]
@kkm000 To debug this, you'll need to attach to the runner as it works. Can you see if you can repro it using the vstest.console.exe bootstrapper instead of from the IDE? It should be running the same child processes and code, but there could be some differences.

There are docs on debugging the adapter here:
https://github.com/xunit/visualstudio.xunit/tree/master/test.harness

**Comment by kkm000** [comment]
@onovotny: sure, will try tonight.

**Comment by kkm000** [comment]
@onovotny: I was banging my head against this for a while, but I cannot reproduce the issue in Visual Studio any more. NUnit runner is also not hanging. I am still trying, but I should take my "can reliably reproduce" back... I do not think I even updated my system, but I rebooted a couple times since. Oh boy, how elusive is this one!

**Comment by darrelmiller** [comment]
Just as a data point, I had this problem with tests hanging when I did Run All using VS2015 Update 1 CTP on Win 10 10565.  I switched the Default Processor Architecture to x64 and the tests run just fine.

**Comment by JranZu** [comment]
Thanks, switching to x64 worked for me. Windows 8.2

**Comment by TheFastCat** [comment]
My unit tests still hang do due apparent threading issues (seems like a deadlock) when running either "Run All Tests" from within Visual Studio and intermittently when selecting a large group of tests, right-clicking and selecting "Run Selected Tests".

I've created a repository on Github for Visual Studio 2013 as demonstration

     https://github.com/TheFastCat/XunitHangRepro

To repro: clone, open, build and then "Run All" from the VS Test Explorer.

Machine is a Microsoft Surface Pro 3, Core i-7 8 gigs of RAM but this seems to happen on Azure Build Controllers as well when executing tests on CI.

One theory - the linked repro is a skinnier version of a fatter code base with a heavy "TestFixtureStartUp" (ala NUnit terms) with more DI, dependent assemblies etc - wondering if the startup and teardown might therefore be a culprit - as this skinner version of the repro executes more tests than the fatter, real version does before hanging.

**Comment by TheFastCat** [comment]
I resolved these error by refactoring the following asynchronous code away from Task.Result and to async/await : 

            "When a POST request is made with the address to add".
                f(() =>
                {
                    Request.Method = HttpMethod.Post;
                    Request.RequestUri = _addressesUri;
                    Request.Content = new ObjectContent<dynamic>(addressToAdd, new JsonMediaTypeFormatter());
                    Response = Client.SendAsync(Request).Result;
                });

to

            "When a POST request is made with the address to add".
                f(async () =>
                {
                    Request.Method = HttpMethod.Post;
                    Request.RequestUri = _addressesUri;
                    Request.Content = new ObjectContent<dynamic>(addressToAdd, new JsonMediaTypeFormatter());
                    Response = await Client.SendAsync(Request);
                });

(and other places) - now the tests don't cause a VS hang.

Lesson learned: avoid Task.Result like the plague!

**Comment by bradwilson** [comment]
> Lesson learned: avoid Task.Result like the plague!

We use VS2015 code analyzers that I wrote to prohibit the use of `.Result`. :)

**Comment by jessegood** [comment]
FWIW after upgrading to [VS2015](https://www.visualstudio.com/en-us/news/vs2015-update1-vs.aspx) I can now use `x86` as default processor architecture.

**Comment by TheFastCat** [comment]
Our build server can't find the tests now - probably due to the architecture issue. It's on Visual Studio Online (TFS) - Is there a way to specify a runsettings (https://goo.gl/2X41Wl) file for Visual Studio to enable the requisite variable to X64?

**Comment by WaffleSquirrel** [comment]
@TheFastCat I had a similar on one of our build servers. Do you know if the VSTest executable is actually executing the unit tests on the build server rather than the default MSTest executable? Running unit tests from Visual Studio 2015 uses vstest.console.exe. A build server would normally use MSTest. 

Our devs on my team would have unit test run and pass fine on their local machine. Then after checking in, some tests would fail while going thru our Go pipeline. The cause for this fo us was that some unit test classes used the MS Fakes framework and relied on discovery of extra generated dlls for the fakes. MSTest was being run on oir build server and it was not able to discover them. I had to install VSTest on the build server and update our PowerShell scripts to run the tests via VSTest.

If the above alone doesn't solve your issue, then it may be related to Environment variables on the build server. Build tasks/definitions and other settings files may use these environment variables to resolve folder paths relating to the installation locations of .Net Framework and its tools etc. I had to set different env vars on our servers that our Go agents ran on since the account running the tests was a local service account. That also solved some other tests that wouldn't work with NUnit and XUnit test adapter dll discovery.

Hope that info can help navigate you to a solution.

**Comment by Akelitz** [comment]
I have the same problem trying to run some xUnit tests both in VS 2015 enterprise and via "dnx test". In VS 2015 I can successfully run the same tests in smaller groups, but I cannot "run all" or run larger group. Please note those tests are migrated from MSTest on which I can run them without any problem.

**Comment by onovotny** [comment]
Are you using VS 2015 Update 1?

**Comment by Akelitz** [comment]
Yes I am.

**Comment by onovotny** [comment]
To be clear is this with dnx tests or desktop?

**Comment by Akelitz** [comment]
DNX

**Comment by onovotny** [comment]
Just trying to narrow things down; does it work from the command line?

**Comment by Akelitz** [comment]
I run "dnx test" from the command line and it hangs the same way

**Comment by onovotny** [comment]
Is there a repro project you can share? VS is essentially calling the command line and interpreting the results, so it should run on the cmd line successfully.

**Comment by Akelitz** [comment]
I'll try to setup something for you. Gimme some time.

**Comment by Akelitz** [comment]
Hi Oren, trying to set up a repro project... I fixed the issue. It was related to my incorrect implementation of Task.FromException (not available on DNX). Sorry for bothering you, may xUnit be with you :)

**Comment by TheFastCat** [comment]
@WaffleSquirrel Thank you so much for your helpful comment (and moreso edit thereof!) -- those resources were just what I needed. Tests and CI now working as intended!

**Comment by bradwilson** [comment]
Closing this, because it appears everybody who had issues is now working correctly.

**Comment by JustinMinnaar** [comment]
Having had a similiar problem, when running all tests hung for a bit, then stopped and showed no tests were run, I narrowed down to the failing test using a binary search (half/half) and found the tested item had a stack overflow exception. Seems this causes the problem, so resolving the stack overflow problem is the first step.

**Comment by rostyslavskarbek** [comment]
Replacing _**Task.Result**_ with **_async/await_** helped to avoid VS Test Runner to hang.
Actually it worked well till I have **_Task.Result_** only in one MyClassTests.cs, but when .Result is used in multiple test lasses then it might cause the issue with VS Test Runner.
@TheFastCat gave a good hint at it.

**Comment by bradwilson** [comment]
> Replacing Task.Result with async/await helped to avoid VS Test Runner to hang.

Use of `Task.Result` should be a huge warning sign, as it can and will cause deadlocks in some scenarios.

**Comment by MaurGi** [comment]
Same behavior with VS 2017 for me - 
I used both the ReSharper runner and the Visual Studio runner - 

.Result is necessary in my case where I call a series of I/O operations in the back end and then wait for their completion - not much of an option for me here.

**Comment by bradwilson** [comment]
The only ways out of this deadlock situation are:

* Stop calling things that deadlock (obviously);
* Turn off thread limitations which cause you to run out of threads.

For the latter, add a [JSON configuration file](https://xunit.github.io/docs/configuring-with-json.html) and set `maxParallelThreads` to `-1`.

**Comment by MaurGi** [comment]
I cannot control if my dependencies are thread safe and the problem might be there - 

On my side, I have removed all my calls to Task.* except a call to Task.Delay(10) to keep my code async - it still deadlocks. 
I don't have any apparent deadlocking-risk code myself as I am firing web requests and waiting for the .Results in the same order I fire them using WaitAll().

Unit tests have no problems running, but when I run my smoke/integration tests it hangs -
Smoke tests do hit the backend and use the full stack, including several NuGet packages (some less common than others..)

Is there a way to run XUnit tests serially with ReSharper in Visual Studio 2017 or with the Visual Studio runner?
I saw [this](https://xunit.github.io/docs/running-tests-in-parallel.html)  but I am not sure how to configure that with the ReSharp / Visual Studio runner.

thx
